src/object.c

```diff
@@ -1247,7 +1247,6 @@ struct redisMemOverhead *getMemoryOverheadData(void) {
     for (j = 0; j < server.dbnum; j++) {
         redisDb *db = server.db+j;
         unsigned long long keyscount = kvstoreSize(db->keys);
-        if (keyscount == 0) continue;
```

Comment:
We currently count and display all databases, but if a user has configured a large number of databases, like 1024, then even if only one db is in use, it would output 1024 results, which is overwhelming and not conducive to observation.

I think we should use a different condition to avoid displaying empty databases, such as skipping them if both the keys and expires have a `kvstoreMemUsage` of zero (also, `kvstoreMemUsage` should no longer include the `sizeof(*kvs)`, as this part is actually already accounted for in `server.initial_memory_usage`).

Reply:
I'm feeling a bit conflicted now. Based on the current method of calculating `kvstoreMemUsage`, even if the dict empty, the memory occupied by `kvstore` itself would still be considered part of the overhead. Therefore, we should return the overhead details for all databases, which means there would be no need for a continue condition. It seems this would be more reasonable?

Reply:
i think it's pointless to list the overhead for all empty databases (ones with no dict buckets at all).
i'd argue that i still rather count that memory in the global overhead metric, but listing these databases is plain noise IMHO.

Reply:
I don't want to list all databases neither, as too many empty dbs would be meaningless.

But currently, we create kvstores with `KVSTORE_ALLOCATE_DICTS_ON_DEMAND`, but without `KVSTORE_FREE_EMPTY_DICTS`. Therefore, after deleting all data (with scan+del or flushdb sync), even if the bucket count is zero, the dict data structure still exists and occupies memory.

This means that a kvstore that has once had data inserted and then cleared will use more memory than a kvstore that has never been used, precisely sizeof(dict)*allocated_dicts more. I think it would be more reasonable to use `allocated_dicts` as a condition here.

Reply:
so listing all the options from low to high:
1. list dbs only if they have actual keys in them
2. list dbs only if they have allocated LUT buckets (what we have now)
3. list dbs only if they have dicts that are non-NULL
4. list dbs only if they use memory, basically listing all

i'm ok with 3. you prefer 4, i'm good with that too.

Reply:
no no, I also prefer 3, just like that:

```diff
--- a/src/kvstore.c
+++ b/src/kvstore.c
@@ -532,6 +533,10 @@ int kvstoreNumNonEmptyDicts(kvstore *kvs) {
     return kvs->non_empty_dicts;
 }

+int kvstoreNumAllocatedDicts(kvstore *kvs) {
+    return kvs->allocated_dicts;
+}
+
 int kvstoreNumDicts(kvstore *kvs) {
     return kvs->num_dicts;
 }

--- a/src/kvstore.h
+++ b/src/kvstore.h
@@ -30,6 +30,7 @@ void kvstoreGetStats(kvstore *kvs, char *buf, size_t bufsize, int full);
 int kvstoreFindDictIndexByKeyIndex(kvstore *kvs, unsigned long target);
 int kvstoreGetNextNonEmptyDictIndex(kvstore *kvs, int didx);
 int kvstoreNumNonEmptyDicts(kvstore *kvs);
+int kvstoreNumAllocatedDicts(kvstore *kvs);
 int kvstoreNumDicts(kvstore *kvs);
 uint64_t kvstoreGetHash(kvstore *kvs, const void *key);

--- a/src/object.c
+++ b/src/object.c
@@ -1246,7 +1246,7 @@ struct redisMemOverhead *getMemoryOverheadData(void) {

     for (j = 0; j < server.dbnum; j++) {
         redisDb *db = server.db+j;
-        if (!kvstoreBuckets(db->keys)) continue;
+        if (!kvstoreNumAllocatedDicts(db->keys)) continue;

         unsigned long long keyscount = kvstoreSize(db->keys);
```

Reply:
sorry, i messed up the counting. i  meant to say that i prefer 2 but i'm ok with 3.
anyway, @CharlesChen888 please make that change and let's move on.